### PR TITLE
Prevent accidental cluster deletion

### PR DESCRIPTION
**Pull request by ironcladlou**
Before this commit, the hosted cluster controller treated a missing
hostedcluster the same as a deleted one. This is incorrect, because any cache or
other transient weirdness from the apiserver for a hostedcluster which actually
still exists will result in that cluster being accidentally destroyed. Absence
of the hostedcluster being reconciled isn't the same as deletion intent.

This commit fixes the problem by short-circuiting reconciliation if the
requested hostedcluster doesn't exist. If on a subsequent event the resource can
be fetched, we'll reconcile it as usual. If we continue to get queued events for
a resource the apiserver says isn't found, that could be some other cache or
connectivity issue we want to surface and resolve. What we should not do is
infer a deletion event from it.

### CODE REVIEW

**Comment by ironcladlou** [comment]
I'm observing this right now. My hypershift-operator is misidentifying a non-deleted hostedcluster as "missing" and so deleted its nodepools and service endpoints, and only stopped short of deleting the entire control plane namespace because an empty HostedCluster struct is fed through to `delete()` causing an HCP lookup failure with an empty name. Something must be very wrong here...

**Comment by ironcladlou** [comment]
Manifests as:

```
{"level":"error","ts":1642194257.0969057,"logger":"controller.hostedcluster","msg":"Reconciler error","reconciler group":"hypershift.openshift.io","reconciler kind":"HostedCluster","name":"dmace-hyper-4e6d","namespace":"clusters","error":"failed to delete cluster: error deleting HostedControlPlane \"\" in namespace \"clusters-dmace-hyper-4e6d\": resource name may not be empty","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/dmace/Projects/hypershift/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/Users/dmace/Projects/hypershift/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227"}
```

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 1]
```diff
@@ -213,19 +213,13 @@ func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 
 	// Look up the HostedCluster instance to reconcile
 	hcluster := &hyperv1.HostedCluster{}
-	isMissing := false
 	err := r.Get(ctx, req.NamespacedName, hcluster)
 	if err != nil {
-		if apierrors.IsNotFound(err) {
-			isMissing = true
```
Comment (sjenning) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:220]: I think we could get in an infinite loop here at if the resource is deleted (NotFound), we return an error, which will requeue the reconcile.  I think we still need to check for `apierrors.IsNotFound(err)` and return with no error in that case.

Comment (ironcladlou) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:220]: I think you're right — I can instead log and return no error.

Comment (ironcladlou) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:220]: Fixed, PTAL

**Comment by ironcladlou** [comment]
To be clear, there's still a bug to run down even with this change. Whatever is happening to cause the client to return notfound is unresolved, but now it'll result in the cluster not being reconciled, instead of the cluster being inappropriately deleted.

**Pull request merged**